From 0b71f561680d62c7851f014be35ff17e2ad3937e Mon Sep 17 00:00:00 2001
From: Andreas Scherer <andreas_scherer@alice.de>
Date: Tue, 1 Sep 2015 18:39:23 +0200
Subject: [PATCH 4/6] Fix intermediate bugs.

Private communication reveals that my patch for mmix-sim.w was wrong.
However, it uncovered a genuine bug at the same line. mmix-sim worked
different than mmix-pipe, although the intention was the same. This
earned me my second MMIX T-shirt since 2000.

The second change was a natural.
---
 abstime.w  | 2 +-
 mmix-sim.w | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/abstime.w b/abstime.w
index 50d6aa9..6605ba1 100644
--- a/abstime.w
+++ b/abstime.w
@@ -18,7 +18,7 @@ hold more than 32 bits.
 #include <stdio.h>
 #include <time.h>
 @#
-main()
+int main()
 {
   printf("#define ABSTIME %ld\n",time(NULL));
   return 0;
diff --git a/mmix-sim.w b/mmix-sim.w
index cb6995c..6d0c0fd 100644
--- a/mmix-sim.w
+++ b/mmix-sim.w
@@ -2657,7 +2657,7 @@ if (sclock.l || sclock.h || !resuming) {
   if ((!(loc.h&sign_bit)||(g[rU].h&0x8000)) &&@|
     ((op&(g[rU].h>>16))==(g[rU].h>>24))) {
       g[rU].l++;
-      if (g[rU].l==0)@+{@+g[rU].h++;@+if (g[rU].h&0x7fff==0) g[rU].h-=0x8000;@+}
+      if (g[rU].l==0)@+{@+g[rU].h++;@+if ((g[rU].h&0x7fff)==0) g[rU].h-=0x8000;@+}
   } /* usage counter counts matched instructions simulated */
   if (g[rI].l<=info[op].oops && g[rI].l && g[rI].h==0) tracing=breakpoint=true;
   g[rI]=incr(g[rI],-info[op].oops); /* interval $\upsilon$ timer counts down */
-- 
2.9.2

